text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/MiniMessageParser.java

```diff
@@ -0,0 +1,517 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.minimessage;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.KeybindComponent;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.TranslatableComponent;
+import net.kyori.adventure.text.event.ClickEvent;
+import net.kyori.adventure.text.event.HoverEvent;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextColor;
+import net.kyori.adventure.text.minimessage.fancy.Fancy;
+import net.kyori.adventure.text.minimessage.fancy.Gradient;
+import net.kyori.adventure.text.minimessage.fancy.Rainbow;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static net.kyori.adventure.text.minimessage.Constants.CLICK;
+import static net.kyori.adventure.text.minimessage.Constants.CLOSE_TAG;
+import static net.kyori.adventure.text.minimessage.Constants.COLOR;
+import static net.kyori.adventure.text.minimessage.Constants.GRADIENT;
+import static net.kyori.adventure.text.minimessage.Constants.HOVER;
+import static net.kyori.adventure.text.minimessage.Constants.INSERTION;
+import static net.kyori.adventure.text.minimessage.Constants.KEYBIND;
+import static net.kyori.adventure.text.minimessage.Constants.PRE;
+import static net.kyori.adventure.text.minimessage.Constants.RAINBOW;
+import static net.kyori.adventure.text.minimessage.Constants.RESET;
+import static net.kyori.adventure.text.minimessage.Constants.SEPARATOR;
+import static net.kyori.adventure.text.minimessage.Constants.TAG_END;
+import static net.kyori.adventure.text.minimessage.Constants.TAG_START;
+import static net.kyori.adventure.text.minimessage.Constants.TRANSLATABLE;
+
+public class MiniMessageParser {
+
+  // regex group names
+  private static final String START = "start";
+  private static final String TOKEN = "token";
+  private static final String INNER = "inner";
+  private static final String END = "end";
+  // https://regex101.com/r/8VZ7uA/10
+  private static final Pattern pattern = Pattern.compile("((?<start><)(?<token>[^<>]+(:(?<inner>['\"]?([^'\"](\\\\['\"])?)+['\"]?))*)(?<end>>))+?");
+
+  private static final Pattern dumSplitPattern = Pattern.compile("['\"]:['\"]");
+
+  private static final Map<Class<? extends Fancy>, Fancy> empty = new HashMap<>();
```

Comment:
this is never mutated, its just a empty constant map, used to avoid an endless recursion in the `applyFormatting` method (and to avoid an object creation there)

Reply:
maybe use `Collections.emptyMap()` then?